File: Week_02/Seminar/TK/find_the_difference/solution_counting.cpp
#include <string>

class Solution
{
public:
    char findTheDifference(std::string s, std::string t)
    {
        const int M = 'z' - 'a' + 1;
        short counts[M]{};

        for (short i = 0; i < s.size(); i++)
        {
            counts[s[i] - 'a']++;
        }

        for (short i = 0; i < t.size(); i++)
        {
            counts[t[i] - 'a']--;
        }

        for (short i = 0; i < M; i++)
        {
            if (counts[i] < 0)
            {
                return 'a' + i;
            }
        }

        return '0';
    }
};

File: Week_08/Seminar/Solutions/group_anagrams/solution_counting.py
class Solution:
    def get_hash(self, word):
        counts = [0] * 26

        for ch in word:
            counts[ord(ch) - ord('a')] += 1
        
        return tuple(counts)

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = defaultdict(list)

        for word in strs:
            word_hash = self.get_hash(word)

            d[word_hash].append(word)
        
        return d.values()
    